src/control_filters/gravity_compensation.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2023, Stogl Robotics Consulting UG (haftungsbeschränkt)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "control_filters/gravity_compensation.hpp"
+
+#include "geometry_msgs/msg/vector3_stamped.hpp"
+#include "geometry_msgs/msg/wrench_stamped.hpp"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include "tf2/LinearMath/Vector3.h"
+
+namespace control_filters
+{
+template <>
+bool GravityCompensation<geometry_msgs::msg::WrenchStamped>::update(
+  const geometry_msgs::msg::WrenchStamped & data_in, geometry_msgs::msg::WrenchStamped & data_out)
+{
+  if (!this->configured_)
+  {
+    RCLCPP_ERROR_SKIPFIRST_THROTTLE((*logger_), *clock_, 2000, "Filter is not configured");
+    return false;
+  }
+
+  // Update internal parameters if required
+  if (parameter_handler_->is_old(parameters_))
+  {
+    parameters_ = parameter_handler_->get_params();
+    compute_internal_params();
+  }
+
+  try
+  {
+    // transform from data_in frame to sensor_frame (most likely to be identity)
+    transform_sensor_datain_ = p_tf_Buffer_->lookupTransform(parameters_.sensor_frame,
+      data_in.header.frame_id, rclcpp::Time());
+
+    // use data_out frame id for the back transformation, otherwise same is data_in
+    if (!data_out.header.frame_id.empty())
+    {
+      data_out.header.stamp = data_in.header.stamp;  // only copy the timestamp
+    }
+    else
+    {
+      data_out.header = data_in.header;  // keep the same header and same frame_id
+    }
+    transform_data_out_sensor_ = p_tf_Buffer_->lookupTransform(
+        data_out.header.frame_id, parameters_.sensor_frame, rclcpp::Time());
+    // rotation only (because gravity is a field) from world (gravity frame) to sensor frame
+    rot_sensor_world_ = p_tf_Buffer_->lookupTransform(
```

Comment:
Are you sure this is correct?

Reply:
for the rotation, well, if I use the full transform, the -mg force along z world gets transformed from the world to sensor frame, generating a force matching -mg in the sensor frame, but also the cross product of the relative distance between sensor and world with the -mg, creating a torque. Gravity is not creating such a torque at the CoG. The only cross product is the distance of CoG to center of measurement with the simple gravity force at CoG (but oriented correctly).
 The same calculation is done there https://github.com/ros-controls/ros2_controllers/blob/master/admittance_controller/include/admittance_controller/admittance_rule_impl.hpp#L318